### PR TITLE
proof-producer: add circuit writer

### PR DESCRIPTION
**Pull request by oclaw**
In addition to assignment writer added circuit_writer impl (tested with CI circuit input)

### CODE REVIEW

File: proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp [hunk 1]
```diff
@@ -501,6 +502,25 @@ namespace nil {
...
                     )
                 );
+
+                return true;
+            }
+
+            bool save_circuit_to_file(boost::filesystem::path circuit_file) {
+                BOOST_LOG_TRIVIAL(info) << "Writing circuit to " << circuit_file;
+                if (!constraint_system_) {
+                    BOOST_LOG_TRIVIAL(error) << "No circuit is currently loaded into the Prover";
```
Comment (akokoshn) [inline proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp:512]: `No circuit is currently loaded`

File: proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp [hunk 2]
```diff
@@ -501,6 +502,25 @@ namespace nil {
...
+                    return false;
+                }
+
+                std::ofstream out(circuit_file.string(), std::ios::binary | std::ios::out);
+                if (!out.is_open()) {
+                    BOOST_LOG_TRIVIAL(error) << "failed to open file " << circuit_file;
```
Comment (akokoshn) [inline proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp:518]: `Failed to open file`

File: proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp [hunk 3]
```diff
@@ -501,6 +502,25 @@ namespace nil {
...
+                    return false;
+                }
+
+                circuit_writer<Endianness, BlueprintField> writer(out);
```
Comment (akokoshn) [inline proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp:522]: May be better to use static method

File: proof-producer/bin/proof-producer/include/nil/proof-generator/circuit_writer.hpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+namespace nil {
+    namespace proof_generator {
+
+
+        template <typename Endianness, typename BlueprintField> 
+        class circuit_writer {
+                public:
+                    using ArithmetizationType = nil::crypto3::zk::snark::plonk_constraint_system<BlueprintField>;
+                    using TTypeBase = nil::marshalling::field_type<Endianness>;
+                    using Circuit = nil::blueprint::circuit<ArithmetizationType>;
```
Comment (akokoshn) [inline proof-producer/bin/proof-producer/include/nil/proof-generator/circuit_writer.hpp:39]: Let's use `nil::crypto3::zk::snark::plonk_constraint_system<BlueprintField>`

**Review by akokoshn — COMMENTED** [review]
Let's make it library, like here: https://github.com/NilFoundation/placeholder/pull/91

File: proof-producer/bin/proof-producer/include/nil/proof-generator/circuit_writer.hpp (hunk 4 above)
Comment (oclaw) [reply proof-producer/bin/proof-producer/include/nil/proof-generator/circuit_writer.hpp:39]: It is already used under `ArithmetizationType`  alias

File: proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp (hunk 1 above)
Comment (oclaw) [reply proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp]: fixed

File: proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp (hunk 2 above)
Comment (oclaw) [reply proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp]: fixed

File: proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp (hunk 3 above)
Comment (oclaw) [reply proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp]: done

File: proof-producer/bin/proof-producer/include/nil/proof-generator/circuit_writer.hpp (hunk 4 above)
Comment (akokoshn) [reply proof-producer/bin/proof-producer/include/nil/proof-generator/circuit_writer.hpp]: i mean use the same type as in prover: https://github.com/NilFoundation/placeholder/blob/b640712c1f95dac05151431376a502b3613c08d9/proof-producer/bin/proof-producer/include/nil/proof-generator/prover.hpp#L158

Comment (oclaw) [reply proof-producer/bin/proof-producer/include/nil/proof-generator/circuit_writer.hpp]: You mean rename ArithmetizationType --> ConstraintSystem?

Comment (oclaw) [reply proof-producer/bin/proof-producer/include/nil/proof-generator/circuit_writer.hpp]: Oh, got it, will fix

Comment (oclaw) [reply proof-producer/bin/proof-producer/include/nil/proof-generator/circuit_writer.hpp]: done

**Review by akokoshn — APPROVED** [review]

**Review by akokoshn — APPROVED** [review]

**Pull request merged**